File: LinkedList.cpp
#include <iostream>
using namespace std;

class Node {
  public:
    int value;
    Node* next;
    Node(int val) {
        value = val;
        next = nullptr;
    }
};

class LinkedList {
  private:
    Node* head;
    Node* tail;
    int length;

  public:
    /* Constructor */
    LinkedList() {
        head = nullptr;
        tail = nullptr;
        length = 0;
    }

    LinkedList(int value) {
        Node* newNode = new Node(value);
        head = newNode;
        tail = newNode;
        length = 1;
    }

    /* Destructor */
    ~LinkedList() {
        Node* temp = head;
        while (temp != nullptr) {
            head = head->next;
            delete temp;
            temp = head;
        }
    }

    /* Methods */
    void printList() {
        if (head == nullptr) {
            cout << "Empty Linked List" << endl;
            return;
        }

        Node* currentNode = head;
        while (currentNode != nullptr) {
            cout << currentNode->value << " ";
            currentNode = currentNode->next;
        }
        cout << endl;
    }

    void getHead() {
        if (head != nullptr) {
            cout << "Head: " << head->value << endl;
            return;
        }
        cout << "Head: nullptr." << endl;
    }

    void getTail() {
        if (tail != nullptr) {
            cout << "Tail: " << tail->value << endl;
            return;
        }
        cout << "Tail: nullptr." << endl;
    }

    void getLength() {
        cout << "Length: " << length << endl;
    }

    bool append(int value) {
        Node* newNode = new Node(value);
        if (head == nullptr) {
            head = newNode;
            tail = newNode;
            length++;
            return true;
        }

        tail->next = newNode;
        tail = newNode;
        length++;
        return true;
    }

    void deleteLast() {
        if (head == nullptr) return;
        if (length == 1) {
            delete head;
            head = nullptr;
            tail = nullptr;
            length--;
            return;
        }

        Node* temp = head;
        Node* prev = nullptr;
        while (temp->next != nullptr) {
            prev = temp;
            temp = temp->next;
        }

        tail = prev;
        tail->next = nullptr;
        length--;
        delete temp;
    }

    bool prepend(int value) {
        Node* newNode = new Node(value);
        if (head == nullptr) {
            head = newNode;
            tail = newNode;
            length++;
            return true;
        }

        newNode->next = head;
        head = newNode;
        length++;
        return true;
    }

    void deleteFirst() {
        if (head == nullptr) return;
        if (length == 1) {
            delete head;
            head = nullptr;
            tail = nullptr;
            length--;
            return;
        }

        Node* temp = head;
        head = head->next;
        temp->next = nullptr;
        length--;
        delete temp;
    }

    Node* get(int index) {
        if (index < 0 || index >= length) return nullptr;
        Node* temp = head;
        for (int i = 0; i < index; i++) {
            temp = temp->next;
        }
        return temp;
    }

    bool set(int index, int value) {
        Node* temp = get(index);
        if (temp == nullptr) return false;
        temp->value = value;
        return true;
    }

    bool insert(int index, int value) {
        if (index < 0 || index > length) return false;
        if (index == 0) return prepend(value);
        if (index == length) return append(value);

        Node* newNode = new Node(value);
        Node* prev = get(index - 1);
        newNode->next = prev->next;
        prev->next = newNode;
        length++;
        return true;
    }

    void deleteNode(int index) {
        if (index < 0 || index >= length) return;
        if (index == 0) return deleteFirst();
        if (index == length - 1) return deleteLast();

        Node* prev = get(index - 1);
        Node* temp = prev->next;
        prev->next = temp->next;
        temp->next = nullptr;
        length--;
        delete temp;
    }

    void reverse() {
        Node* temp = head;
        head = tail;
        tail = temp;

        Node* after = temp->next;
        Node* before = nullptr;
        for (int i = 0; i < length; i++) {
            after = temp->next;
            temp->next = before;
            before = temp;
            temp = after;
        }
    }
};

int main() {
    LinkedList* myLinkedList = new LinkedList(1);

    myLinkedList->append(2);
    myLinkedList->append(3);

    myLinkedList->deleteLast();
    myLinkedList->deleteLast();
    myLinkedList->deleteLast();

    myLinkedList->prepend(4);
    myLinkedList->prepend(5);

    // myLinkedList->deleteFirst();
    // myLinkedList->set(0, 88);

    myLinkedList->insert(2, 99);
    myLinkedList->insert(1, 77);

    myLinkedList->getHead();
    myLinkedList->getTail();
    myLinkedList->getLength();
    myLinkedList->printList();

    myLinkedList->reverse();
    myLinkedList->printList();

    delete myLinkedList;
    return 0;
}

File: LinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self, value=None):
        if value is None:
            self.head = None
            self.tail = None
            self.length = 0
            return

        new_node = Node(value)
        self.head = new_node
        self.tail = new_node
        self.length = 1

    def print_list(self):
        if self.head is None:
            print("Empty Linked List")
            return

        temp = self.head
        while temp is not None:
            print(temp.value)
            temp = temp.next

    def append(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
            self.length += 1
            return True

        self.tail.next = new_node
        self.tail = new_node
        self.length += 1
        return True

    def pop(self):
        if self.head is None:
            return None

        temp = self.head
        if self.length == 1:
            self.head = None
            self.tail = None
            self.length -= 1
            return temp

        prev = self.head
        while temp.next:
            prev = temp
            temp = temp.next
        self.tail = prev
        self.tail.next = None
        self.length -= 1
        return temp

    def prepend(self, value):
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
            self.length += 1
            return True

        new_node.next = self.head
        self.head = new_node
        self.length += 1
        return True

    def pop_first(self):
        if self.head is None:
            return None

        temp = self.head
        if self.length == 1:
            self.head = None
            self.tail = None
            self.length -= 1
            return temp

        self.head = self.head.next
        temp.next = None
        self.length -= 1
        return temp

    def get(self, index):
        if index < 0 or index >= self.length:
            return None

        temp = self.head
        for _ in range(index):
            temp = temp.next
        return temp

    def set_value(self, index, value):
        temp = self.get(index)
        if temp is None:
            return False

        temp.value = value
        return True

    def insert(self, index, value):
        if index < 0 or index > self.length:
            return False

        if index == 0:
            return self.prepend(value)

        if index == self.length:
            return self.append(value)

        new_node = Node(value)
        prev = self.get(index - 1)
        new_node.next = prev.next
        prev.next = new_node
        self.length += 1
        return True

    def remove(self, index):
        if index < 0 or index >= self.length:
            return None

        if index == 0:
            return self.pop_first()

        if index == self.length - 1:
            return self.pop()

        prev = self.get(index - 1)
        temp = prev.next
        prev.next = temp.next
        temp.next = None
        self.length -= 1
        return temp

    def reverse(self):
        temp = self.head
        self.head = self.tail
        self.tail = temp

        after = temp.next
        before = None
        for _ in range(self.length):
            after = temp.next
            temp.next = before
            before = temp
            temp = after


my_linked_list = LinkedList(4)
my_linked_list.append(5)
my_linked_list.append(6)

my_linked_list.pop()
my_linked_list.pop()
my_linked_list.pop()
my_linked_list.pop()

my_linked_list.prepend(1)
my_linked_list.prepend(2)
my_linked_list.prepend(3)

my_linked_list.set_value(1, 99)
my_linked_list.insert(2, 55)
my_linked_list.print_list()

my_linked_list.reverse()
my_linked_list.print_list()
